### PR TITLE
joplin: new, 2.1.9

### PR DESCRIPTION
**Pull request by Fearyncess**
Topic Description
-----------------
Add `joplin`, which is a opensource note taking app.

Package(s) Affected
-------------------
* `joplin`

Security Update?
----------------
No

Architectural Progress
----------------------

- [x] AMD64 `amd64`   
- [x] AArch64 `arm64`

----

After the pull request is merged, all package(s) affected must be rebuilt against the `stable` Git tree and environment (only `stable` repository should be enabled in `sources.list`). This section marks the progress above.

Please, make sure the list of architectures below matches the ones above.

Post-Merge Architectural Progress
---------------------------------

- [ ] AMD64 `amd64`   
- [ ] AArch64 `arm64`

### CODE REVIEW

File: extra-productivity/joplin/autobuild/build [hunk 1]
```diff
@@ -0,0 +1,46 @@
+abinfo "Start building Joplin..."
+cd "${SRCDIR}"
+
+abinfo "Building Joplin modules and Downloading dependencies..."
+npm install --unsafe-perm --cache ${_NPM_CACHE}
+
+
```
Comment (MingcongBai) [inline extra-productivity/joplin/autobuild/build:8]: Extra empty line here.

File: extra-productivity/joplin/autobuild/build [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+abinfo "Building Joplin CLI..."
+cd "${SRCDIR}/packages/app-cli"
+npm run build
+cd "${SRCDIR}/packages/app-cli/build"
+_CLI_PACK=$(npm pack | tail -n 1)
+npm install --production --unsafe-perm --global --user root --cache ${_NPM_CACHE} --prefix "${PKGDIR}/usr" "${SRCDIR}/packages/app-cli/build/${_CLI_PACK}"
```
Comment (MingcongBai) [inline extra-productivity/joplin/autobuild/build]: Probably a good idea to wrap your lines with `\`.

File: extra-productivity/joplin/autobuild/build [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+cd "${PKGDIR}/usr"
+mv "${PKGDIR}/usr/lib/node_modules/joplin" "${PKGDIR}/usr/lib/joplin"
+rm -rv "${PKGDIR}/usr/lib/node_modules"
+
+abinfo "Fixing npm permissions problem for Joplin CLI..."
+find "${PKGDIR}/usr" -type d -exec chmod 755 {} +
```
Comment (MingcongBai) [inline extra-productivity/joplin/autobuild/build]: Missing verbosity.

File: extra-productivity/joplin/autobuild/build [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+abinfo "Building Joplin CLI..."
+cd "${SRCDIR}/packages/app-cli"
+npm run build
+cd "${SRCDIR}/packages/app-cli/build"
+_CLI_PACK=$(npm pack | tail -n 1)
+npm install --production --unsafe-perm --global --user root --cache ${_NPM_CACHE} --prefix "${PKGDIR}/usr" "${SRCDIR}/packages/app-cli/build/${_CLI_PACK}"
+
+cd "${PKGDIR}/usr"
+mv "${PKGDIR}/usr/lib/node_modules/joplin" "${PKGDIR}/usr/lib/joplin"
```
Comment (MingcongBai) [inline extra-productivity/joplin/autobuild/build]: Missing verbosity.

File: extra-productivity/joplin/autobuild/build [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+abinfo "Fixing npm permissions problem for Joplin CLI..."
+find "${PKGDIR}/usr" -type d -exec chmod 755 {} +
+
+abinfo "Remove unneeded dependencies info in package.json..."
+find "${PKGDIR}" -name package.json -print0 | xargs -0 sed -i "/_where/d"
+jq '.|=with_entries(select(.key|test("_.+")|not))' "${PKGDIR}/usr/lib/joplin/package.json" > "${PKGDIR}/usr/lib/joplin/package.json"
+chmod 644 "${PKGDIR}/usr/lib/joplin/package.json"
+
+abinfo "Make all .so library files executable for AutoBuild3 QA checks..."
+find "${PKGDIR}/usr/lib/joplin/node_modules" -name "*.so.*" -exec chmod +x {} +
```
Comment (MingcongBai) [inline extra-productivity/joplin/autobuild/build]: Missing verbosity.

File: extra-productivity/joplin/autobuild/build [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+mv "${PKGDIR}/usr/lib/node_modules/joplin" "${PKGDIR}/usr/lib/joplin"
+rm -rv "${PKGDIR}/usr/lib/node_modules"
+
+abinfo "Fixing npm permissions problem for Joplin CLI..."
+find "${PKGDIR}/usr" -type d -exec chmod 755 {} +
+
+abinfo "Remove unneeded dependencies info in package.json..."
+find "${PKGDIR}" -name package.json -print0 | xargs -0 sed -i "/_where/d"
+jq '.|=with_entries(select(.key|test("_.+")|not))' "${PKGDIR}/usr/lib/joplin/package.json" > "${PKGDIR}/usr/lib/joplin/package.json"
+chmod 644 "${PKGDIR}/usr/lib/joplin/package.json"
```
Comment (MingcongBai) [inline extra-productivity/joplin/autobuild/build]: Missing verbosity.

File: extra-productivity/joplin/autobuild/build [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+abinfo "Make all .so library files executable for AutoBuild3 QA checks..."
+find "${PKGDIR}/usr/lib/joplin/node_modules" -name "*.so.*" -exec chmod +x {} +
+
+
+abinfo "Building Joplin Desktop..."
+cd "${SRCDIR}/packages/app-desktop"
+mkdir -p "${SRCDIR}/packages/app-desktop/dist/"
```
Comment (MingcongBai) [inline extra-productivity/joplin/autobuild/build]: Missing verbosity.

File: extra-productivity/joplin/autobuild/build [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+touch "${SRCDIR}/packages/app-desktop/dist/AppImage"
+USE_HARD_LINKS=false npm run dist -- --publish=never --linux "${ARCH_OPTS}" --dir="${SRCDIR}/packages/app-desktop/dist/"
+mv "${SRCDIR}/packages/app-desktop/dist/${ARCH_DIR}/@joplinapp-desktop" "${SRCDIR}/packages/app-desktop/dist/${ARCH_DIR}/joplin-desktop"
```
Comment (MingcongBai) [inline extra-productivity/joplin/autobuild/build]: Missing verbosity.

File: extra-productivity/joplin/autobuild/build [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+abinfo "Remove duplicated and unpacked electron application files..."
+rm -rv "${SRCDIR}/packages/app-desktop/dist/${ARCH_DIR}/resources/app.asar.unpacked"
+cp -a "${SRCDIR}/packages/app-desktop/dist/${ARCH_DIR}" "${PKGDIR}/usr/lib/joplin-desktop"
```
Comment (MingcongBai) [inline extra-productivity/joplin/autobuild/build]: Missing verbosity.

File: extra-productivity/joplin/autobuild/build [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+abinfo "Remove unneeded dependencies info in package.json..."
+find "${PKGDIR}" -name package.json -print0 | xargs -0 sed -i "/_where/d"
+jq '.|=with_entries(select(.key|test("_.+")|not))' "${PKGDIR}/usr/lib/joplin/package.json" > "${PKGDIR}/usr/lib/joplin/package.json"
+chmod 644 "${PKGDIR}/usr/lib/joplin/package.json"
+
+abinfo "Make all .so library files executable for AutoBuild3 QA checks..."
+find "${PKGDIR}/usr/lib/joplin/node_modules" -name "*.so.*" -exec chmod +x {} +
+
+
```
Comment (MingcongBai) [inline extra-productivity/joplin/autobuild/build:36]: Extra blank line.

File: extra-productivity/joplin/autobuild/build [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+mkdir -p "${SRCDIR}/packages/app-desktop/dist/"
+touch "${SRCDIR}/packages/app-desktop/dist/AppImage"
+USE_HARD_LINKS=false npm run dist -- --publish=never --linux "${ARCH_OPTS}" --dir="${SRCDIR}/packages/app-desktop/dist/"
+mv "${SRCDIR}/packages/app-desktop/dist/${ARCH_DIR}/@joplinapp-desktop" "${SRCDIR}/packages/app-desktop/dist/${ARCH_DIR}/joplin-desktop"
+
+abinfo "Remove duplicated and unpacked electron application files..."
+rm -rv "${SRCDIR}/packages/app-desktop/dist/${ARCH_DIR}/resources/app.asar.unpacked"
+cp -a "${SRCDIR}/packages/app-desktop/dist/${ARCH_DIR}" "${PKGDIR}/usr/lib/joplin-desktop"
+
+
```
Comment (MingcongBai) [inline extra-productivity/joplin/autobuild/build:47]: Extra blank line.

File: extra-productivity/joplin/autobuild/overrides/usr/share/applications/joplin-desktop.desktop [hunk 12]
```diff
@@ -0,0 +1,9 @@
+[Desktop Entry]
+Name=Joplin
+Comment=Joplin is an open source note-taking app. Capture your thoughts and securely access them from any device.
```
Comment (MingcongBai) [inline extra-productivity/joplin/autobuild/overrides/usr/share/applications/joplin-desktop.desktop]: Application *, do not abbreviate desktop applications.

File: extra-productivity/joplin/autobuild/overrides/usr/share/applications/joplin-desktop.desktop [hunk 13]
```diff
@@ -0,0 +1,9 @@
...
+Exec=/usr/bin/joplin-desktop
+Icon=/usr/lib/joplin-desktop/resources/build/icons/128x128.png
+Terminal=false
+Type=Application
+Categories=Application;Office;
+
```
Comment (MingcongBai) [inline extra-productivity/joplin/autobuild/overrides/usr/share/applications/joplin-desktop.desktop]: Trailing blank line here.

File: extra-productivity/joplin/autobuild/prepare [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+abinfo "Setting the target architecture..."
+if [[ "${CROSS:-$ARCH}" = "amd64" ]]; then
+    export ARCH_OPTS="--x64"
+    export ARCH_DIR="linux-unpacked"
+elif [[ "${CROSS:-$ARCH}" = "arm64" ]]; then
+    export ARCH_OPTS="--arm64"
+    export ARCH_DIR="linux-arm64-unpacked"
+fi
+
+
```
Comment (MingcongBai) [inline extra-productivity/joplin/autobuild/prepare]: Extra empty line here.

File: extra-productivity/joplin/autobuild/prepare [hunk 15]
```diff
@@ -0,0 +1,22 @@
+abinfo "Setting Locale Environment Variable for npm..."
+export LANG=en_US.UTF-8
```
Comment (MingcongBai) [inline extra-productivity/joplin/autobuild/prepare]: Is this really necessary?

File: extra-productivity/joplin/spec [hunk 16]
```diff
@@ -0,0 +1,4 @@
+VER=2.1.9
+SRCS="tbl::https://github.com/laurent22/joplin/archive/refs/tags/v${VER}.tar.gz"
+CHKSUMS="sha256::d175f6c2daabf90264b4e6a0f083153a5a82a512d533fbf1a4bb6915da51d6a1"
+CHKUPDATE="github::repo=laurent22/joplin;pattern=v(\d+\.\d+\.\d+)"
```
Comment (MingcongBai) [inline extra-productivity/joplin/spec]: Use `anitya::id=227031` instead.

**Review by MingcongBai — CHANGES_REQUESTED** [review]

File: extra-productivity/joplin/autobuild/prepare (hunk 15 above)
Comment (Fearyncess) [reply extra-productivity/joplin/autobuild/prepare]: Yes. 

Ref: https://github.com/alfredopalhares/joplin-pkgbuild/issues/25

Comment (MingcongBai) [reply extra-productivity/joplin/autobuild/prepare]: Well we default to C.UTF-8.

Comment (Fearyncess) [reply extra-productivity/joplin/autobuild/prepare]: tested. this tweak doesn't need

**Comment by Fearyncess** [comment]
image
looks good on AMD64

requests ARM64 test

**Comment by OriginCode** [comment]
LGTM on arm64 (PBP):
image

**Review by OriginCode — APPROVED** [review]

**Review by MingcongBai — APPROVED** [review]

**Comment by Cyanoxygen** [comment]
LGTM on Chocobo (my PBP):

image

**Comment by Cyanoxygen** [comment]
Oh too late sorry

**Comment by peeweep** [comment]
Thanks, I finally saw this. I've been looking forward to using joplin desktop before.

**Pull request merged**